src/mgmt/srv.c

```diff
@@ -59,23 +60,104 @@ static struct daos_rpc_handler mgmt_handlers[] = {
 
 #undef X
 
+static void
+process_killrank_request(Drpc__Call *drpc_req, Proto__DaosResponse *daos_resp)
+{
+	Proto__DaosRank	*pb_rank = NULL;
+
+	proto__daos_response__init(daos_resp);
+
+	/* Unpack the daos request from the drpc call body */
+	pb_rank = proto__daos_rank__unpack(
+		NULL, drpc_req->body.len, drpc_req->body.data);
+
+	if (pb_rank == NULL) {
+		daos_resp->status = PROTO__DAOS_REQUEST_STATUS__ERR_UNKNOWN;
+		D_ERROR("Failed to extract rank from request\n");
+
+		return;
+	}
+
+	/* response status is populated with SUCCESS on init */
+	D_DEBUG(DB_MGMT, "Received request to kill rank (%u) on pool (%s)\n",
+		pb_rank->rank, pb_rank->pool_uuid);
+
+	/* TODO: do something with request and populate daos response status */
+
+	proto__daos_rank__free_unpacked(pb_rank, NULL);
+}
+
+static void
+process_drpc_request(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
+{
+	Proto__DaosResponse	*daos_resp = NULL;
+	uint8_t			*body = NULL;
+	size_t			len = 0;
+
+	D_ALLOC_PTR(daos_resp);
+	if (daos_resp == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Failed to allocate daos response ref\n");
+
+		return;
+	}
+
+	switch (drpc_req->method) {
+	case DRPC_METHOD_MGMT_KILL_RANK:
+		/**
+		 * Process daos request and populate daos response,
+		 * command errors should be indicated inside daos response.
+		 */
+		process_killrank_request(drpc_req, daos_resp);
+
+		len = proto__daos_response__get_packed_size(daos_resp);
+		D_ALLOC(body, len);
+		if (body == NULL) {
+			drpc_resp->status = DRPC__STATUS__FAILURE;
+			D_ERROR("Failed to allocate drpc response body\n");
+
+			break;
+		}
+
+		if (proto__daos_response__pack(daos_resp, body) != len) {
+			drpc_resp->status = DRPC__STATUS__FAILURE;
+			D_ERROR("Unexpected num bytes for daos resp\n");
+
+			break;
+		}
+
+		/* Populate drpc response body with packed daos response */
+		drpc_resp->body.len = len;
+		drpc_resp->body.data = body;
+
+		break;
+	default:
+		drpc_resp->status = DRPC__STATUS__UNKNOWN_METHOD;
+		D_ERROR("Unknown method\n");
+	}
+
+	D_FREE(daos_resp);
+}
+
 static void
 mgmt_drpc_handler(Drpc__Call *request, Drpc__Response **response)
 {
-	Drpc__Response *resp;
+	Drpc__Response		*drpc_resp = NULL;
+
+	D_ALLOC_PTR(drpc_resp);
+	if (drpc_resp == NULL) {
+		D_ERROR("Failed to allocate drpc response ref\n");
 
-	D_ALLOC_PTR(resp);
-	if (resp == NULL) {
-		D_ERROR("Failed to allocate response\n");
 		return;
 	}
 
-	drpc__response__init(resp);
-	resp->sequence = request->sequence;
-	resp->status = DRPC__STATUS__SUCCESS;
+	drpc__response__init(drpc_resp);
```

Comment:
we should probably export rpc_reply_create under the name drpc_reply_create() to hide all of this.

Reply:
where would that live? presumably not in this file

Reply:
I'd suggest src/common/drpc.c, since it seems like generic logic.

Reply:
I don't think it's necessary to do as part of this PR though. IMO it would make sense to create a separate ticket/PR to do the drpc API changes. I'd be willing to pick that up.

Reply:
Agreed. Let's do it in a follow-on PR. Li Wei told me that he has some patches that would like to land before doing those changes.